File: probs/1463.py
X = int(input())

dp = [1000000 for _ in range(X)] + [0]

if X == 1:
    print(0)
else:
    while X > 1:
        if X % 3 == 0:
            dp[X // 3] = min(dp[X // 3], dp[X] + 1)
        if X % 2 == 0:
            dp[X // 2] = min(dp[X // 2], dp[X] + 1)
        if X > 1:
            dp[X - 1] = min(dp[X - 1], dp[X] + 1)
        X -= 1
    print(dp[1])

File: probs/1463.cpp
#include <iostream>
#include <string.h>

#define LEN 1000000
int dp[LEN];

using namespace std;

int main()
{
    int N;
    memset(dp, LEN, LEN * sizeof(int));
    cin >> N;
    dp[N] = 0;

    while (N) {
        if (N % 3 == 0) {
            dp[N/3] = min(dp[N] + 1, dp[N/3]);
        }
        if (N % 2 == 0) {
            dp[N/2] = min(dp[N] + 1, dp[N/2]);
        }
        dp[N-1] = min(dp[N] + 1, dp[N-1]);
        N--;
    }
    cout << dp[1] << endl;

    return 0;
}
